### PR TITLE
 Don't use AnnotationBuilder.fromName in DefaultQualifierPolymorphism

### PR DESCRIPTION
**Pull request by mernst**

### CODE REVIEW

**Comment by mernst** [comment]
Merge after #2254

File: framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java [hunk 1]
```diff
@@ -43,17 +43,19 @@ public DefaultQualifierPolymorphism(ProcessingEnvironment env, AnnotatedTypeFact
                 continue;
             }
 
-            Name plval = AnnotationUtils.getElementValueClassName(aa, "value", true);
+            Class<? extends Annotation> plval =
```
[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java]: `fromName` is an inherently error-prone interface.  It must never be called with any annotation class that has elements/fields, but there is no way to check that.  If the contract is violated, then invalid annotations are created, which cause problems elsewhere (such as incorrect equality test results, and crashes in Annotation File Utilities).

`fromClass` is able to check the contract at run time.

I don't believe this is a performance bottleneck, but avoiding errors and malformed data would be worth a small slowdown.

[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java]: As this is the motivation for a few of your PRs, let's discuss it here.
The current implementation of `AnnotationBuilder.fromClass` is:

```java
    public static AnnotationMirror fromClass(
            Elements elements, Class<? extends Annotation> aClass) {
        return fromName(elements, aClass.getCanonicalName());
    }
```

I didn't see a PR that would change that.
So I don't see how one would be inherently better than the other.

[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java]: The eventual goal is to have fromClass check its contract and to make fromName, which cannot check its contract, private.  So many corrupted AnnotationMirrors are created today that this is not possible yet, so it needs to be done in parts.  This is one step toward that goal.

[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java]: Can you describe the contract you want to check in more detail?
`fromName` looks up the `Element` for the annotation: `final TypeElement annoElt = elements.getTypeElement(name);`
At that point we could check what methods the element has - which would correspond to elements/fields in the annotation.
We could probably also check whether the methods have default values and allow no-argument construction of an annotation if all methods have default values.
What contract do you want to enforce over the `Class` that couldn't be equally, even more easily, be enforced over the `TypeElement`?

[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/poly/DefaultQualifierPolymorphism.java]: I don't think that removing all uses of `fromName` is a good goal.
The current uses of `Name` are safe, as they come from the compiler and are not arbitrary strings.
As e.g. #2253 illustrates, rewriting these uses to use `fromClass` results in ugly changes to the API, far offsetting the benefit of having only `fromClass`.
We can add a comment to prefer `fromClass` over `fromName`, but replacing all current uses isn't beneficial.

I found that Issue #722 already exists to track implementing the checks to ensure `build()` and `fromName` verify they have all elements/fields.

**Pull request closed without merging**